sql/analyzer/costed_index_scan.go

```diff
@@ -320,12 +321,13 @@ func getCostedIndexScan(ctx *sql.Context, statsProv sql.StatsProvider, rt sql.Ta
 		bestStat = stats.UpdateCounts(bestStat)
 	}
 
-	if bestStat.FuncDeps().HasMax1Row() && !qFlags.JoinIsSet() && !qFlags.SubqueryIsSet() && lookup.Ranges.Len() == 1 {
-		// Strict index lookup without a join or subquery scope will return
-		// at most one row. We could also use some sort of scope counting
-		// to check for single scope.
-		qFlags.Set(sql.QFlagMax1Row)
-	}
+	// Conservative: disable Max1Row optimization here due to lack of context for SetOp detection (dolt#9641)
```

Comment:
An easier solution would be to unset the QFlagMax1Row in the resolve unions rule:
```diff
diff --git a/sql/analyzer/resolve_unions.go b/sql/analyzer/resolve_unions.go
index 213470932..ea4be8e4d 100644
--- a/sql/analyzer/resolve_unions.go
+++ b/sql/analyzer/resolve_unions.go
@@ -99,6 +99,10 @@ func finalizeUnions(ctx *sql.Context, a *Analyzer, n sql.Node, scope *plan.Scope
                if err != nil {
                        return nil, transform.SameTree, err
                }
+
+               // TODO: comment about why this is needed
+               qFlags.Unset(sql.QFlagMax1Row)
+
                return newN, transform.NewTree, nil
        })
 }
 ```